# [\#11987 Issue](https://github.com/ethereum/solidity/issues/11987) `open`: Error message about disallowed argument location should explain why the location is not allowed
**Labels**: `enhancement`, `help wanted`, `good first issue`, `difficulty: easy`


#### <img src="https://avatars.githubusercontent.com/u/2530770?u=a2b81f85d207864b7db06415db53010c21633b33&v=4" width="50">[nventuro](https://github.com/nventuro) opened issue at [2021-09-16 20:37](https://github.com/ethereum/solidity/issues/11987):

When compiling the following contract with solc v0.8.7:

```
pragma solidity ^0.8.0;

contract Test {
    struct MyStruct {
        uint256 a;
    }
    
    function foo(MyStruct storage bar) public;
}
```

we get the following error message:

```
TypeError: Data location must be "memory" or "calldata" for parameter in function, but "storage" was given.
    function foo(MyStruct storage bar) public;
                 ^------------------^
```

This is not particularly helpful as functions can actually receive `storage` arguments, but only if they are `internal` or `private`. The error message is not particularly insightful on this distinction, nor helps the developer achieve their task.

### Suggestion

A simple improvement could be:
```
Data location must be "memory" or "calldata" for parameter in "public" or "external" function
```

Even better, a suggestion could be appended at the end:

```
TypeError: Data location must be "memory" or "calldata" for parameter in "public" or "external" function, but 
"storage" was given.
    function foo(MyStruct storage bar) public;
                 ^------------------^

Use "internal" or "private" functions to pass storage pointers.
```

#### <img src="https://avatars.githubusercontent.com/u/20340?v=4" width="50">[axic](https://github.com/axic) commented at [2021-09-16 20:37](https://github.com/ethereum/solidity/issues/11987#issuecomment-921236992):

I think this would be as a nice improvement. Did you find similar other cases?

#### <img src="https://avatars.githubusercontent.com/u/2530770?u=a2b81f85d207864b7db06415db53010c21633b33&v=4" width="50">[nventuro](https://github.com/nventuro) commented at [2021-09-16 20:37](https://github.com/ethereum/solidity/issues/11987#issuecomment-921816390):

Hm I don't recall any off any of the top of my head - I just run into this the other day and was puzzled by the error message until I realized what was going on.

#### <img src="https://avatars.githubusercontent.com/u/29651595?u=8046f5d3c40d459521ba2037401d74bd43c1ae40&v=4" width="50">[jhb10c](https://github.com/jhb10c) commented at [2021-09-16 20:37](https://github.com/ethereum/solidity/issues/11987#issuecomment-923279115):

 Hey @cameel, @chriseth. I would like to work on this. I can update the error to the suggestion by @nventuro. If there are any specifications, could you all provide more info?

#### <img src="https://avatars.githubusercontent.com/u/137030?v=4" width="50">[cameel](https://github.com/cameel) commented at [2021-09-16 20:37](https://github.com/ethereum/solidity/issues/11987#issuecomment-923285509):

Well, the version with adjusting the message is straightforward:
- find the message in the code
- tweak description
- run `isoltest` to update tests

A version with the extra suggestion would be harder. You can have an extra message if you specify a secondary location - but that's not what they were meant for and if would be weird if secondary location was pointing at the same thing. But maybe it would make sense if you made it point at the `public`/`external` keyword?

You could also issue two errors, one with "info" severity meant to contain just this extra suggestion - but there's the risk that the user will want to hide info messages.

Neither of these seems particularly nice so I I'd go with just tweaking the text unless you can figure out a nicer solution.

#### <img src="https://avatars.githubusercontent.com/u/29651595?u=8046f5d3c40d459521ba2037401d74bd43c1ae40&v=4" width="50">[jhb10c](https://github.com/jhb10c) commented at [2021-09-16 20:37](https://github.com/ethereum/solidity/issues/11987#issuecomment-923629501):

Thanks for the instructions. I will update the error message for now and leave the extra suggestion for later. There are some similar cases. As such I will also update the error messages for external/public functions with the following data types: bytes; uint[]; structs; mapping(uint => uint); mapping(uint => uint)[].

#### <img src="https://avatars.githubusercontent.com/u/29651595?u=8046f5d3c40d459521ba2037401d74bd43c1ae40&v=4" width="50">[jhb10c](https://github.com/jhb10c) commented at [2021-09-16 20:37](https://github.com/ethereum/solidity/issues/11987#issuecomment-927029869):

Alright, the files are updated and I ran isoltest to check the updated tests. I think everything works fine. 

I will rebase and submit the pull request later today.

#### <img src="https://avatars.githubusercontent.com/u/29651595?u=8046f5d3c40d459521ba2037401d74bd43c1ae40&v=4" width="50">[jhb10c](https://github.com/jhb10c) commented at [2021-09-16 20:37](https://github.com/ethereum/solidity/issues/11987#issuecomment-927206560):

Spoke to soon. There is an issue. I have updated the tests and ran isoltest. However, when I reproduce the error with solc, it produces the original error. I then tried to rebuild the compiler with the updated tests. The original error was still there.  

This leads me to think: 1.) I have not updated the correct test; or 2.) I am building the compiler incorrectly.  @cameel, do you have any advice? Thanks again! 


(The main locations of the error message is dataLocations/publicFunction and types/mapping. The updated files can be found on my branch.)

#### <img src="https://avatars.githubusercontent.com/u/137030?v=4" width="50">[cameel](https://github.com/cameel) commented at [2021-09-16 20:37](https://github.com/ethereum/solidity/issues/11987#issuecomment-927366408):

The same error might be raised from multiple places. Are you sure you found them all? If you are sure you did, I'd suggest manually modifying expectations in all tests that have this message to see which ones fail. Then you can use a debugger to see which path the compiler takes there. If that's still not enough to figure this out, you can also submit a draft PR with your code and these failing tests and we'll take a look.

#### <img src="https://avatars.githubusercontent.com/u/29651595?u=8046f5d3c40d459521ba2037401d74bd43c1ae40&v=4" width="50">[jhb10c](https://github.com/jhb10c) commented at [2021-09-16 20:37](https://github.com/ethereum/solidity/issues/11987#issuecomment-928669242):

Thank you. I appreciate the advice. The message locations have been triple checked and all instances of error message was tracked down. I have submitted a pr.

#### <img src="https://avatars.githubusercontent.com/u/137030?v=4" width="50">[cameel](https://github.com/cameel) commented at [2021-09-16 20:37](https://github.com/ethereum/solidity/issues/11987#issuecomment-937077421):

@jhb10c Do you want to try to add the extra message that suggests to use `internal` or `private`? It can be done using a secondary location, like here:
https://github.com/ethereum/solidity/blob/a9b99dd63b545e5ca8968058cfb8518206b8e343/libsolidity/interface/CompilerStack.cpp#L186-L194

You'd just leave the location field empty and only provide a message.


-------------------------------------------------------------------------------



[Export of Github issue for [ethereum/solidity](https://github.com/ethereum/solidity). Generated on 2022.05.23 at 03:51:38.]
